src/document/json/rht.rs

```diff
@@ -0,0 +1,307 @@
+use crate::document::time::ticket;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+/// RHTNode is a node of RHT(Replicated Hashtable).
+struct RHTNode {
+    key: String,
+    val: String,
+    updated_at: ticket::Ticket,
+    removed_at: Option<ticket::Ticket>,
+}
+
+impl RHTNode {
+    pub fn new(key: String, val: String, updated_at: ticket::Ticket) -> RHTNode {
+        RHTNode {
+            key,
+            val,
+            updated_at,
+            removed_at: None,
+        }
+    }
+
+    /// key returns the key of this node.
+    pub fn key(&self) -> &str {
+        &self.key
+    }
+
+    /// value returns the value of this node.
+    pub fn value(&self) -> &str {
+        &self.val
+    }
+
+    /// updated_at returns the last update time.
+    pub fn updated_at(&self) -> &ticket::Ticket {
+        &self.updated_at
+    }
+
+    /// removed_at returns the deletion time of this node.
+    pub fn removed_at(&self) -> Option<&ticket::Ticket> {
+        if let Some(removed_at) = &self.removed_at {
+            return Some(&removed_at);
+        }
+
+        return None;
+    }
+
+    /// remove removes this node. It only marks the deleted time (tombstone).
+    pub fn remove(&mut self, removed_at: ticket::Ticket) {
+        if let Some(v) = &self.removed_at {
+            if removed_at.after(v) {
+                self.removed_at = Some(removed_at)
+            }
+            return;
+        }
+        self.removed_at = Some(removed_at);
+    }
+
+    pub fn is_removed(&self) -> bool {
+        if let None = self.removed_at {
+            return false;
+        }
+        true
+    }
+}
+
+/// RHT is a hashtable with logical clock(Replicated hashtable).
+/// For more details about RHT: http://csl.skku.edu/papers/jpdc11.pdf
+pub struct RHT {
+    node_map_by_key: HashMap<String, Rc<RefCell<RHTNode>>>,
+    node_map_by_created_at: HashMap<String, Rc<RefCell<RHTNode>>>,
```

Comment:
Can we use a general data type for the keys of Map in Rust?

For instance:
```suggestion
    node_map_by_created_at: HashMap<TimeTicket, Rc<RefCell<RHTNode>>>,
```

Reply:
Yes. it is possible.
ref: https://doc.rust-lang.org/rust-by-example/std/hash/alt_key_types.html

I'll try this :)

Reply:
Yes. it is possible.
ref: https://doc.rust-lang.org/rust-by-example/std/hash/alt_key_types.html

I'll try this :)